// totem/lib/service/auth.js
// Generated by CoffeeScript 1.6.3
(function() {
  var AuthService, Observable, hapilib;

  hapilib = require("hapilib/lib/solo/index");

  Observable = require("../mixin/observable");

  AuthService = (function() {
    $.extend(AuthService.prototype, Observable);

    function AuthService(hapilib) {
      var _this = this;
      this.hapilib = hapilib;
      if (this.hapilib == null) {
        this.hapilib = hapilib;
      }
      this.hapilib.onSessionExpired(function() {
        return _this.notify("sessionExpired");
      });
      this.notifiableEvents = ["sessionExpired"];
    }

    AuthService.prototype.authenticate = function(type, data, callback) {
      type = type.charAt(0).toUpperCase() + type.slice(1);
      return this["authenticate" + type](data, callback);
    };

    AuthService.prototype.authenticateClient = function(data, callback) {
      var _this = this;
      return this.login(false, callback, data, function(data) {
        return _this.hapilib.authenticateClient(callback);
      });
    };

    AuthService.prototype.authenticateFacebook = function(data, callback) {
      var _this = this;
      return this.login(true, callback, data, function(data) {
        return _this.hapilib.facebookLogin(data.signedRequest, callback);
      });
    };

    AuthService.prototype.authenticateCode = function(data, callback) {
      var _this = this;
      return this.login(true, callback, data, function(data) {
        return _this.hapilib.serverExchangeCode(data.code, callback);
      });
    };

    AuthService.prototype.logout = function() {
      return this.hapilib.logout();
    };

    AuthService.prototype.login = function(requireUser, callback, data, authStrategy) {
      var _this = this;
      return this.hapilib.autoLogin(function(err) {
        var _ref;
        if (err || (requireUser && (((_ref = hapilib.user) != null ? _ref.Id : void 0) == null))) {
          return _this.doLogin(authStrategy, data, callback);
        } else {
          return typeof callback === "function" ? callback() : void 0;
        }
      });
    };

    AuthService.prototype.doLogin = function(authStrategy, data, callback) {
      if (typeof data === "function") {
        return data(function(_data) {
          return authStrategy(_data, callback);
        });
      } else {
        return authStrategy(data, callback);
      }
    };

    return AuthService;

  })();

  if (typeof module !== "undefined" && module !== null) {
    module.exports = AuthService;
  }

}).call(this);
