// totem/lib/service/purchase.js
// Generated by CoffeeScript 1.6.3
(function() {
  var PurchaseService;

  PurchaseService = (function() {
    function PurchaseService() {}

    PurchaseService.prototype.purchase = function(params) {
      var allGoods, confStrategy, d, e, good, goodId, payload, paymentService, user, userGoods,
        _this = this;
      good = params.good, allGoods = params.allGoods, userGoods = params.userGoods, user = params.user, paymentService = params.paymentService, confStrategy = params.confStrategy;
      if (user.attributes.Id == null) {
        throw new Error("PurchaseService: User has no Id");
      }
      d = $.Deferred();
      if (confStrategy == null) {
        confStrategy = function(user, goodId, paymentResponse) {
          return _this.confirmPurchase(user, goodId, paymentResponse);
        };
      }
      payload = {
        response: null,
        error: null
      };
      try {
        user.uncache("goods");
        goodId = this.goodId(good, allGoods);
        this.goodIsOwned(good, userGoods);
        paymentService.purchase(user.attributes.Id, goodId, function(err, res) {
          if (err) {
            payload.error = err;
            return d.resolve(payload);
          }
          payload.response = res;
          return confStrategy(user, goodId, res).then(function(err) {
            payload.error = err;
            return d.resolve(payload);
          });
        });
      } catch (_error) {
        e = _error;
        payload.error = e;
        return d.resolve(payload);
      }
      return d.promise();
    };

    PurchaseService.prototype.confirmPurchase = function(user, goodId, paymentResponse) {
      return user.goods().then(function(goods) {
        var good, _i, _len;
        for (_i = 0, _len = goods.length; _i < _len; _i++) {
          good = goods[_i];
          if (good.Id === goodId) {
            return null;
          }
        }
        return new Error("PurchaseService: purchase couldn't be confirmed");
      });
    };

    PurchaseService.prototype.goodId = function(good, goods) {
      var g, _i, _len;
      for (_i = 0, _len = goods.length; _i < _len; _i++) {
        g = goods[_i];
        if (g.Label === good) {
          return g.Id;
        }
      }
      throw new Error("PurchaseService: no such good " + good);
    };

    PurchaseService.prototype.goodIsOwned = function(good, userGoods) {
      var g, _i, _len, _results;
      _results = [];
      for (_i = 0, _len = userGoods.length; _i < _len; _i++) {
        g = userGoods[_i];
        if (g.Label === good) {
          throw new Error("PurchaseService: user owns good " + good);
        }
      }
      return _results;
    };

    return PurchaseService;

  })();

  if (typeof module !== "undefined" && module !== null) {
    module.exports = PurchaseService;
  }

}).call(this);
