// hapilib/lib/com/huzutech/hapi/rest/HAPIRest.js
// Generated by CoffeeScript 1.6.3
(function() {
  var HAPIErrorParser, HAPIModule, HAPIRest, RequestManager,
    __hasProp = {}.hasOwnProperty,
    __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; };

  RequestManager = require("./RequestManager");

  HAPIModule = require("../objects/HAPIModule");

  HAPIErrorParser = require("./HAPIErrorParser");

  HAPIRest = (function(_super) {
    __extends(HAPIRest, _super);

    HAPIRest.include(HAPIErrorParser);

    function HAPIRest() {
      this.headers = {
        "Accept": "application/json"
      };
      this.baseUrl = "";
      this.requestManager = new RequestManager;
    }

    HAPIRest.prototype.abortAll = function() {
      return this.requestManager.abortAll();
    };

    HAPIRest.prototype.requestCount = function() {
      return this.requestManager.count();
    };

    HAPIRest.prototype.setBaseUrl = function(baseUrl) {
      this.baseUrl = baseUrl;
    };

    HAPIRest.prototype.setCustomHeaders = function(headers) {
      var key, value, _results;
      _results = [];
      for (key in headers) {
        if (!__hasProp.call(headers, key)) continue;
        value = headers[key];
        _results.push(this.headers[key] = value);
      }
      return _results;
    };

    HAPIRest.prototype.removeCustomHeader = function(header) {
      return delete this.headers[header];
    };

    HAPIRest.prototype.get = function(url, callback) {
      return this.makeRestCall({
        url: url,
        method: "GET"
      }, callback);
    };

    HAPIRest.prototype["delete"] = function(url, callback) {
      return this.makeRestCall({
        url: url,
        method: "DELETE"
      }, callback);
    };

    HAPIRest.prototype.put = function(url, data, callback) {
      return this.makeDataRequest({
        url: url,
        method: "PUT",
        data: data
      }, callback);
    };

    HAPIRest.prototype.post = function(url, data, callback) {
      return this.makeDataRequest({
        url: url,
        method: "POST",
        data: data
      }, callback);
    };

    HAPIRest.prototype.onError = function(callback) {
      if (this.globalErrorCallbacks == null) {
        this.globalErrorCallbacks = [];
      }
      if (callback != null) {
        return this.globalErrorCallbacks.push(callback);
      }
    };

    HAPIRest.prototype.removeOnError = function(callback) {
      var backup, cb, _i, _len, _ref;
      if (callback == null) {
        return;
      }
      backup = [];
      _ref = this.globalErrorCallbacks;
      for (_i = 0, _len = _ref.length; _i < _len; _i++) {
        cb = _ref[_i];
        if (cb !== callback) {
          backup.push;
        }
      }
      return this.globalErrorCallbacks = backup;
    };

    HAPIRest.prototype.makeRequest = function(params, callback) {
      throw new Error("" + this.constructor.name + "::makeRequest - Override me!!");
    };

    HAPIRest.prototype.notifyError = function(err) {
      var callback, _i, _len, _ref, _results;
      if (this.globalErrorCallbacks != null) {
        _ref = this.globalErrorCallbacks;
        _results = [];
        for (_i = 0, _len = _ref.length; _i < _len; _i++) {
          callback = _ref[_i];
          _results.push(callback(err));
        }
        return _results;
      }
    };

    HAPIRest.prototype.callbackValidation = function(callback) {
      if (!((callback != null) && callback instanceof Function)) {
        throw new Error("callback function not provided");
      }
    };

    HAPIRest.prototype.makeDataRequest = function(params, callback) {
      var heads, key, value, _ref;
      heads = {};
      _ref = this.headers;
      for (key in _ref) {
        if (!__hasProp.call(_ref, key)) continue;
        value = _ref[key];
        heads[key] = value;
      }
      heads["Content-Type"] = "application/json";
      params.headers = heads;
      return this.makeRestCall(params, callback);
    };

    HAPIRest.prototype.makeRestCall = function(params, callback) {
      this.callbackValidation(callback);
      if (!params.headers) {
        params.headers = this.headers;
      }
      return this.makeRequest({
        url: this.baseUrl + params.url,
        method: params.method,
        headers: params.headers,
        data: params.data != null ? JSON.stringify(params.data) : void 0
      }, callback);
    };

    return HAPIRest;

  })(HAPIModule);

  if (typeof module !== "undefined" && module !== null) {
    module.exports = HAPIRest;
  }

}).call(this);
