// hapilib/lib/com/huzutech/hapi/rest/RequestManager.js
// Generated by CoffeeScript 1.6.3
(function() {
  var RequestManager,
    __hasProp = {}.hasOwnProperty;

  RequestManager = (function() {
    function RequestManager() {
      this.requests = {};
    }

    RequestManager.prototype.count = function() {
      var count, k, v, _ref;
      count = 0;
      _ref = this.requests;
      for (k in _ref) {
        if (!__hasProp.call(_ref, k)) continue;
        v = _ref[k];
        count++;
      }
      return count;
    };

    RequestManager.prototype.add = function(req) {
      if (req == null) {
        throw new Error(this.constructor.Name + "::add -  Not a valid request");
      }
      if ((req.requestId != null) && (this.requests[req.requestId] != null)) {
        return;
      }
      req.requestId = (Math.random()).toString(36).slice(2);
      return this.requests[req.requestId] = req;
    };

    RequestManager.prototype.remove = function(req) {
      if (req == null) {
        return;
      }
      return delete this.requests[req.requestId];
    };

    RequestManager.prototype.abortAll = function() {
      var k, v, _ref, _results;
      _ref = this.requests;
      _results = [];
      for (k in _ref) {
        if (!__hasProp.call(_ref, k)) continue;
        v = _ref[k];
        v.abort();
        _results.push(this.remove(v));
      }
      return _results;
    };

    return RequestManager;

  })();

  if (typeof module !== "undefined" && module !== null) {
    module.exports = RequestManager;
  }

}).call(this);
