// hapilib/lib/com/huzutech/hapi/rest/TokenManager.js
// Generated by CoffeeScript 1.6.3
(function() {
  var HAPIRest, TokenManager;

  HAPIRest = require("./HAPIRest");

  TokenManager = (function() {
    function TokenManager(restClient) {
      var _this = this;
      this.restClient = restClient;
      if (!((this.restClient != null) && this.restClient instanceof HAPIRest)) {
        this.restClient = void 0;
        throw new TypeError("You need to pass an implementation of HAPIRest");
      }
      this.restClient.onError(function(err) {
        if (err.statusCode === 401 && err.message === "accessToken is invalid. token has expired") {
          _this.removeToken();
          return typeof _this.tokenExpiredCallback === "function" ? _this.tokenExpiredCallback() : void 0;
        }
      });
    }

    TokenManager.prototype.setToken = function(token, expiresIn) {
      var _this = this;
      this.token = token;
      if (expiresIn == null) {
        throw new Error("expiresIn param required");
      }
      this.restClient.setCustomHeaders({
        Authorization: "Bearer " + this.token
      });
      return this.tokenExpirationTimeout = setTimeout(function() {
        _this.removeToken();
        return typeof _this.tokenExpiredCallback === "function" ? _this.tokenExpiredCallback() : void 0;
      }, expiresIn * 1000);
    };

    TokenManager.prototype.removeToken = function() {
      if (this.tokenExpirationTimeout != null) {
        clearTimeout(this.tokenExpirationTimeout);
      }
      this.restClient.removeCustomHeader("Authorization");
      return this.token = null;
    };

    TokenManager.prototype.onTokenExpired = function(tokenExpiredCallback) {
      this.tokenExpiredCallback = tokenExpiredCallback;
    };

    return TokenManager;

  })();

  if (typeof module !== "undefined" && module !== null) {
    module.exports = TokenManager;
  }

}).call(this);
