// totem/lib/util/promise.js
// Generated by CoffeeScript 1.6.3
(function() {
  var Promise;

  Promise = {
    queue: function(name, element, task) {
      var d;
      d = $.Deferred();
      $(element).queue(name, function(next) {
        return task().then(function() {
          d.resolve();
          next();
          if (($(element).queue(name)).length === 0) {
            return $(element).data("dequeuing", false);
          }
        });
      });
      if (!$(element).data("dequeuing")) {
        $(element).data("dequeuing", true);
        $(element).dequeue(name);
      }
      return d.promise();
    },
    sequence: function(tasks) {
      var i, seq, task, _i, _len;
      seq = $.Deferred().resolve();
      if (typeof (tasks != null ? tasks[0] : void 0) !== "function") {
        return seq;
      }
      for (i = _i = 0, _len = tasks.length; _i < _len; i = ++_i) {
        task = tasks[i];
        seq = (function(i) {
          return seq.then(function() {
            return tasks[i]();
          });
        })(i);
      }
      return seq;
    },
    heap: function() {
      return {
        done: function() {
          var _ref;
          if (!(((_ref = this.heap) != null ? _ref.length : void 0) > 0)) {
            return $.Deferred().resolve();
          }
          return $.when.apply($, this.heap);
        },
        add: function(task) {
          if (this.heap == null) {
            this.heap = [];
          }
          return this.heap.push(task());
        },
        clear: function() {
          this.heap = [];
          return this.done();
        }
      };
    }
  };

  if (typeof module !== "undefined" && module !== null) {
    module.exports = Promise;
  }

}).call(this);
