// totem/lib/user/authorized.js
// Generated by CoffeeScript 1.6.3
(function() {
  var AbstractUser, AuthorizedUser, Lang,
    __hasProp = {}.hasOwnProperty,
    __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; };

  AbstractUser = require("./abstract");

  Lang = require("../util/lang");

  AuthorizedUser = (function(_super) {
    __extends(AuthorizedUser, _super);

    function AuthorizedUser(client, attributes) {
      this.client = client;
      this.setAttributes(attributes);
      this.cache = {};
    }

    AuthorizedUser.prototype.goods = function() {
      var d,
        _this = this;
      d = $.Deferred();
      if (this.cache.goods != null) {
        return d.resolve(Lang.clone(this.cache.goods));
      }
      this.client.get("/users/" + this.attributes.Id + "/goods", function(err, goods) {
        if (err) {
          return d.reject(err);
        }
        _this.cache.goods = goods;
        return d.resolve(Lang.clone(_this.cache.goods));
      });
      return d.promise();
    };

    AuthorizedUser.prototype.avatars = function(options) {
      var assets, d,
        _this = this;
      if (options == null) {
        options = {};
      }
      d = $.Deferred();
      if (options.assets == null) {
        options.assets = true;
      }
      if ((options.Label != null) || (options.Id != null)) {
        return this.avatar(options);
      }
      assets = options.assets ? "?include=tree" : "";
      this.client.get("/users/" + this.attributes.Id + "/avatars" + assets, function(err, avatars) {
        if (err) {
          return d.reject(err);
        }
        return d.resolve(Lang.clone(avatars));
      });
      return d.promise();
    };

    AuthorizedUser.prototype.avatar = function(options) {
      var assets, d, identifier,
        _this = this;
      identifier = options.Label || options.Id;
      if (!identifier) {
        throw new Error("User::avatar needs a label or id");
      }
      d = $.Deferred();
      assets = options.assets ? "?include=tree" : "";
      this.client.get("/avatars/" + identifier + assets, function(err, avatar) {
        if (err) {
          return d.reject(err);
        }
        return d.resolve(Lang.clone(avatar));
      });
      return d.promise();
    };

    AuthorizedUser.prototype.setAttributes = function(attributes) {
      var key, value, _results;
      if (this.attributes == null) {
        this.attributes = {};
      }
      _results = [];
      for (key in attributes) {
        value = attributes[key];
        _results.push(this.attributes[key] = value);
      }
      return _results;
    };

    return AuthorizedUser;

  })(AbstractUser);

  if (typeof module !== "undefined" && module !== null) {
    module.exports = AuthorizedUser;
  }

}).call(this);
