// totem/lib/drawing/adaptor/abstract.js
// Generated by CoffeeScript 1.6.3
(function() {
  var AbstractAdaptor;

  AbstractAdaptor = (function() {
    function AbstractAdaptor() {}

    AbstractAdaptor.prototype.prepareCanvas = function(canvas) {
      if ($(canvas).data("totem-prepared")) {
        return canvas;
      }
      if (!$(canvas).is(this.elementType)) {
        canvas = this.getElementFrom(canvas);
      }
      $(canvas).data("totem-prepared", true);
      return canvas;
    };

    AbstractAdaptor.prototype.getElementFrom = function(canvas) {
      var element;
      element = $(canvas).children(this.elementType);
      if (element.length > 0) {
        return element.get(0);
      }
      return this.createElementFrom(canvas);
    };

    AbstractAdaptor.prototype.createElementFrom = function(canvas) {
      var element;
      element = $("<" + this.elementType + ">");
      element = this.styleElementFrom(canvas, element);
      $(canvas).append(element);
      return element.get(0);
    };

    AbstractAdaptor.prototype.styleElementFrom = function(canvas, element) {
      $(element).width($(canvas).width());
      $(element).height($(canvas).height());
      return element;
    };

    AbstractAdaptor.prototype.loadImage = function(image) {
      var d, imageElement;
      d = $.Deferred();
      if (image instanceof Image) {
        if (this.imageComplete(image)) {
          d.resolve(image);
        } else {
          image.onload = function() {
            return d.resolve(image);
          };
        }
      } else {
        imageElement = new Image;
        imageElement.onload = function() {
          return d.resolve(imageElement);
        };
        imageElement.onerror = function() {
          return d.resolve(null);
        };
        imageElement.src = image;
      }
      return d.promise();
    };

    AbstractAdaptor.prototype.imageComplete = function(image) {
      return image.complete || image.width + image.height > 0;
    };

    return AbstractAdaptor;

  })();

  if (typeof module !== "undefined" && module !== null) {
    module.exports = AbstractAdaptor;
  }

}).call(this);
