// totem/lib/drawing/adaptor/canvasElement.js
// Generated by CoffeeScript 1.6.3
(function() {
  var AbstractAdaptor, Canvas, CanvasElement, Promise, _ref,
    __hasProp = {}.hasOwnProperty,
    __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; };

  Promise = require("../../util/promise");

  AbstractAdaptor = require("./abstract");

  Canvas = require("./mixin/canvas");

  CanvasElement = (function(_super) {
    __extends(CanvasElement, _super);

    function CanvasElement() {
      _ref = CanvasElement.__super__.constructor.apply(this, arguments);
      return _ref;
    }

    $.extend(CanvasElement.prototype, Canvas);

    CanvasElement.prototype.queue = "totemDrawingQueue";

    CanvasElement.prototype.elementType = "canvas";

    CanvasElement.prototype.drawOn = function(composite, canvas) {
      var category, name, stack, tasks, _fn, _i, _len,
        _this = this;
      tasks = [
        function() {
          _this.clearCanvas(canvas);
          return $.Deferred().resolve();
        }
      ];
      for (_i = 0, _len = composite.length; _i < _len; _i++) {
        category = composite[_i];
        _fn = function(stack) {
          return tasks.push(function() {
            return _this.drawStackOn(stack, canvas);
          });
        };
        for (name in category) {
          stack = category[name];
          _fn(stack);
        }
      }
      return Promise.queue(this.queue, canvas, function() {
        return Promise.sequence(tasks);
      });
    };

    CanvasElement.prototype.clearCanvas = function(canvas) {
      var context;
      canvas = this.prepareCanvas(canvas);
      context = canvas.getContext("2d");
      return context.clearRect(0, 0, canvas.width, canvas.height);
    };

    CanvasElement.prototype.drawStackOn = function(stack, canvas) {
      var cell, tasks, _fn, _i, _len,
        _this = this;
      tasks = [];
      _fn = function(cell) {
        return tasks.push(function() {
          return _this.drawCellOn(cell, canvas);
        });
      };
      for (_i = 0, _len = stack.length; _i < _len; _i++) {
        cell = stack[_i];
        _fn(cell);
      }
      return Promise.sequence(tasks);
    };

    CanvasElement.prototype.drawCellOn = function(cell, canvas) {
      var _this = this;
      canvas = this.prepareCanvas(canvas);
      return this.loadImage(cell.image).then(function(image) {
        var colour, dimensions;
        if (!image) {
          return;
        }
        if ((colour = cell.colour)) {
          dimensions = {
            width: canvas.width,
            height: canvas.height
          };
          image = _this.colourise(image, colour, dimensions);
        }
        return canvas.getContext("2d").drawImage(image, 0, 0);
      });
    };

    CanvasElement.prototype.styleElementFrom = function(canvas, element) {
      element.attr("width", $(canvas).width());
      element.attr("height", $(canvas).height());
      return element;
    };

    return CanvasElement;

  })(AbstractAdaptor);

  if (typeof module !== "undefined" && module !== null) {
    module.exports = CanvasElement;
  }

}).call(this);
