// totem/lib/drawing/adaptor/colourisableUListElement.js
// Generated by CoffeeScript 1.6.3
(function() {
  var Canvas, ColourisableUListElement, UListElement, _ref,
    __hasProp = {}.hasOwnProperty,
    __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; };

  UListElement = require("./uListElement");

  Canvas = require("./mixin/canvas");

  ColourisableUListElement = (function(_super) {
    __extends(ColourisableUListElement, _super);

    function ColourisableUListElement() {
      _ref = ColourisableUListElement.__super__.constructor.apply(this, arguments);
      return _ref;
    }

    $.extend(ColourisableUListElement.prototype, Canvas);

    ColourisableUListElement.prototype.createCell = function(cell, canvas) {
      var li;
      li = ColourisableUListElement.__super__.createCell.call(this, cell, canvas);
      if (cell.colour != null) {
        return this.getColourisedCell(li, cell);
      }
      return li;
    };

    ColourisableUListElement.prototype.getColourisedCell = function(li, cell) {
      var colour,
        _this = this;
      colour = cell.colour;
      this.heap.add(function() {
        return _this.loadImage(cell.image).then(function(image) {
          var canvas, dimensions;
          dimensions = {
            width: li.width(),
            height: li.height()
          };
          canvas = _this.colourise(image, colour, dimensions);
          li.css({
            backgroundImage: "none"
          });
          return li.append(canvas);
        });
      });
      return li;
    };

    return ColourisableUListElement;

  })(UListElement);

  if (typeof module !== "undefined" && module !== null) {
    module.exports = ColourisableUListElement;
  }

}).call(this);
