// totem/lib/drawing/adaptor/uListElement.js
// Generated by CoffeeScript 1.6.3
(function() {
  var AbstractAdaptor, Promise, UListElement,
    __hasProp = {}.hasOwnProperty,
    __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; };

  Promise = require("../../util/promise");

  AbstractAdaptor = require("./abstract");

  UListElement = (function(_super) {
    __extends(UListElement, _super);

    UListElement.prototype.elementType = "ul";

    function UListElement() {
      this.heap = Promise.heap();
    }

    UListElement.prototype.drawOn = function(composite, canvas) {
      var category, name, stack, _i, _len;
      this.clearCanvas(canvas);
      for (_i = 0, _len = composite.length; _i < _len; _i++) {
        category = composite[_i];
        for (name in category) {
          stack = category[name];
          this.drawStackOn(stack, canvas);
        }
      }
      return this.heap.done();
    };

    UListElement.prototype.clearCanvas = function(canvas) {
      this.heap.clear();
      canvas = this.prepareCanvas(canvas);
      return $(canvas).empty();
    };

    UListElement.prototype.drawCellOn = function(cell, canvas) {
      var li;
      canvas = this.prepareCanvas(canvas);
      li = this.createCell(cell, canvas);
      $(canvas).append(li);
      return this.heap.done();
    };

    UListElement.prototype.drawStackOn = function(stack, canvas) {
      var cell, _i, _len;
      for (_i = 0, _len = stack.length; _i < _len; _i++) {
        cell = stack[_i];
        this.drawCellOn(cell, canvas);
      }
      return this.heap.done();
    };

    UListElement.prototype.createCell = function(cell, canvas) {
      var li,
        _this = this;
      this.heap.add(function() {
        return _this.loadImage(cell.image);
      });
      li = $("<li></li>");
      li.css({
        display: "block",
        position: "absolute",
        width: $(canvas).width(),
        height: $(canvas).height(),
        backgroundImage: "url(" + cell.image + ")",
        backgroundPosition: "50% 50%",
        backgroundSize: "contain",
        backgroundRepeat: "no-repeat"
      });
      return li;
    };

    UListElement.prototype.styleElementFrom = function(canvas, element) {
      element = UListElement.__super__.styleElementFrom.call(this, canvas, element);
      element.css({
        listStyle: "none",
        margin: 0,
        padding: 0
      });
      return element;
    };

    return UListElement;

  })(AbstractAdaptor);

  if (typeof module !== "undefined" && module !== null) {
    module.exports = UListElement;
  }

}).call(this);
