// hapilib/lib/com/huzutech/hapi/HAPI.js
// Generated by CoffeeScript 1.6.3
(function() {
  var HAPI, HAPIExchangeCode, HAPIFacebook, HAPIJQuery, HAPIModule, HAPIRest, TokenManager, exportObj,
    __hasProp = {}.hasOwnProperty,
    __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; };

  HAPIModule = require("./objects/HAPIModule");

  HAPIExchangeCode = require("./HAPIExchangeCode");

  HAPIFacebook = require("./HAPIFacebook");

  HAPIRest = require("./rest/HAPIRest");

  HAPIJQuery = require("./rest/HAPIJQuery");

  TokenManager = require("./rest/TokenManager");

  HAPI = (function(_super) {
    __extends(HAPI, _super);

    HAPI.include(HAPIFacebook);

    HAPI.include(HAPIExchangeCode);

    function HAPI(restClient) {
      this.restClient = restClient;
      if (!((this.restClient != null) && this.restClient instanceof HAPIRest)) {
        throw new TypeError("You need to pass an implementation of HAPIRest");
      }
      this.tokenManager = new TokenManager(this.restClient);
    }

    HAPI.prototype.setInstance = function(instanceId) {
      this.instanceId = instanceId;
      if (instanceId != null) {
        return this.restClient.setCustomHeaders({
          InstanceId: this.instanceId
        });
      } else {
        return this.restClient.removeCustomHeader("InstanceId");
      }
    };

    HAPI.prototype.init = function(params) {
      this.endPoint = params.endPoint;
      this.clientId = params.clientId;
      this.instanceId = params.instanceId;
      return this.restClient.setBaseUrl(this.endPoint);
    };

    HAPI.prototype.authenticateClient = function(callback) {
      var _this = this;
      return this.restClient.post("/authenticate/client", {
        clientId: this.clientId
      }, function(err, data) {
        if (err) {
          return callback(err);
        }
        return _this.serverExchangeCode(data.authCode, callback);
      });
    };

    HAPI.prototype.authenticateUser = function(username, password, callback) {
      var _this = this;
      return this.restClient.post("/authenticate/user", {
        ClientId: this.clientId,
        Username: username,
        Password: password,
        InstanceId: this.instanceId
      }, function(err, data) {
        if (err) {
          return callback(err);
        }
        if (data.authCode) {
          return _this.serverExchangeCode(data.authCode, callback);
        } else if (data.accessToken) {
          return _this.completeLoginProcess(data.accessToken, data.expiresIn, callback);
        } else {
          return callback(new Error("Could not retrieve accessToken"));
        }
      });
    };

    HAPI.prototype.authenticateCmsUser = function(username, password, callback) {
      var _this = this;
      return this.restClient.post("/authenticate/cmsuser", {
        Username: username,
        Password: password
      }, function(err, data) {
        if (err) {
          return callback(err);
        }
        return _this.completeCmsLoginProcess(data.accessToken, data.expiresIn, callback);
      });
    };

    HAPI.prototype.authenticateFromCms = function(instanceId, callback) {
      var _this = this;
      this.setInstance(instanceId);
      return this.restClient.post("/cms/instance/token", {
        InstanceId: instanceId
      }, function(err, data) {
        if (err) {
          return callback(err);
        }
        return _this.completeLoginProcess(data.accessToken, data.expiresIn, callback);
      });
    };

    HAPI.prototype.autoLoginCms = function(callback) {
      var storedToken,
        _this = this;
      if ((typeof localStorage !== "undefined" && localStorage !== null ? localStorage.getItem("HAPITokenCms") : void 0) != null) {
        storedToken = localStorage.getItem("HAPITokenCms");
        return this.restClient.get("/cms/tokens/" + (encodeURIComponent(storedToken)), function(err, data) {
          if ((err != null) || data.expiresIn === 0) {
            _this.removeStoredCmsToken();
            return callback(new Error("Stored TokenManager has expired"));
          } else {
            return _this.completeCmsLoginProcess(storedToken, data.expiresIn, callback);
          }
        });
      } else {
        return callback(new Error("Not logged in"));
      }
    };

    HAPI.prototype.autoLogin = function(callback) {
      var storedInstance, storedToken,
        _this = this;
      if (((typeof localStorage !== "undefined" && localStorage !== null ? localStorage.getItem("HAPIToken") : void 0) != null) && ((typeof localStorage !== "undefined" && localStorage !== null ? localStorage.getItem("HAPIInstance") : void 0) != null)) {
        storedInstance = localStorage.getItem("HAPIInstance");
        storedToken = localStorage.getItem("HAPIToken");
        this.setInstance(storedInstance);
        return this.restClient.get("/tokens/" + (encodeURIComponent(storedToken)), function(err, data) {
          if ((err != null) || (data != null ? data.expiresIn : void 0) === 0) {
            _this.removeStoredToken();
            return callback(new Error("Stored TokenManager has expired"));
          } else {
            return _this.completeLoginProcess(storedToken, data.expiresIn, callback);
          }
        });
      } else {
        return callback(new Error("Not logged in"));
      }
    };

    HAPI.prototype.logout = function() {
      this.removeStoredToken();
      this.removeStoredCmsToken();
      this.tokenManager.removeToken();
      return this.user = void 0;
    };

    HAPI.prototype.onSessionExpired = function(onSessionExpiredCallback) {
      this.onSessionExpiredCallback = onSessionExpiredCallback;
      return this.tokenManager.onTokenExpired(this.onSessionExpiredCallback);
    };

    HAPI.prototype.completeCmsLoginProcess = function(accessToken, expiresIn, callback) {
      var _this = this;
      this.tokenManager.setToken(accessToken, expiresIn);
      this.masterAccessToken = accessToken;
      this.masterExpiresIn = expiresIn;
      if (typeof localStorage !== "undefined" && localStorage !== null) {
        localStorage.setItem("HAPITokenCms", accessToken);
      }
      this.setInstance(null);
      return this.restClient.get("/cms/me", function(err, data) {
        _this.user = data;
        return callback(err);
      });
    };

    HAPI.prototype.completeLoginProcess = function(accessToken, expiresIn, callback) {
      var _this = this;
      this.tokenManager.setToken(accessToken, expiresIn);
      if (typeof localStorage !== "undefined" && localStorage !== null) {
        localStorage.setItem("HAPIToken", accessToken);
      }
      return this.restClient.get("/me", function(err, data) {
        if (data == null) {
          data = {
            Username: "guest"
          };
        }
        _this.user = data;
        return callback(err);
      });
    };

    HAPI.prototype.removeStoredToken = function() {
      if ((typeof localStorage !== "undefined" && localStorage !== null ? localStorage.getItem("HAPIToken") : void 0) != null) {
        return localStorage.removeItem("HAPIToken");
      }
    };

    HAPI.prototype.removeStoredCmsToken = function() {
      if ((typeof localStorage !== "undefined" && localStorage !== null ? localStorage.getItem("HAPITokenCms") : void 0) != null) {
        return localStorage.removeItem("HAPITokenCms");
      }
    };

    return HAPI;

  })(HAPIModule);

  exportObj = {
    Singleton: new HAPI(new HAPIJQuery()),
    HAPI: HAPI
  };

  if (typeof module !== "undefined" && module !== null) {
    module.exports = exportObj;
  }

}).call(this);
