// totem/lib/mixin/observable.js
// Generated by CoffeeScript 1.6.3
(function() {
  var Observable,
    __indexOf = [].indexOf || function(item) { for (var i = 0, l = this.length; i < l; i++) { if (i in this && this[i] === item) return i; } return -1; };

  Observable = {
    notifiableEvents: [],
    attachObserver: function(observer, events) {
      var e, _i, _len, _results;
      if (!(events instanceof Array)) {
        events = [events];
      }
      _results = [];
      for (_i = 0, _len = events.length; _i < _len; _i++) {
        e = events[_i];
        if (__indexOf.call(this.notifiableEvents, e) < 0) {
          throw new Error("" + this.constructor + " does not notify the event '" + e + "'.");
        }
        _results.push(this.addObserver(observer, e));
      }
      return _results;
    },
    detachObserver: function(observer, events) {
      var e, _i, _len, _ref, _results;
      if (!(events instanceof Array)) {
        events = [events];
      }
      _results = [];
      for (_i = 0, _len = events.length; _i < _len; _i++) {
        e = events[_i];
        if ((((_ref = this._observers) != null ? _ref[e] : void 0) != null) && __indexOf.call(this._observers[e], observer) >= 0) {
          _results.push(this.removeObserver(observer, this._observers[e]));
        }
      }
      return _results;
    },
    notifyObserver: function(observer, event, data) {
      return observer.notify(event, data);
    },
    notify: function(event, data) {
      var observer, _i, _len, _ref, _ref1, _results;
      if (data == null) {
        data = {};
      }
      if (((_ref = this._observers) != null ? _ref[event] : void 0) == null) {
        return;
      }
      data.sender = this;
      _ref1 = this._observers[event];
      _results = [];
      for (_i = 0, _len = _ref1.length; _i < _len; _i++) {
        observer = _ref1[_i];
        _results.push(this.notifyObserver(observer, event, data));
      }
      return _results;
    },
    addObserver: function(observer, event) {
      var _base;
      if (this._observers == null) {
        this._observers = {};
      }
      if ((_base = this._observers)[event] == null) {
        _base[event] = [];
      }
      if (__indexOf.call(this._observers[event], observer) < 0) {
        return this._observers[event].push(observer);
      }
    },
    removeObserver: function(observer, collection) {
      return collection.splice(collection.indexOf(observer, 1));
    }
  };

  if (typeof module !== "undefined" && module !== null) {
    module.exports = Observable;
  }

}).call(this);
