// totem/lib/avatar.js
// Generated by CoffeeScript 1.6.3
(function() {
  var Avatar, Lang, Observable, WithAuthorization;

  Lang = require("./util/lang");

  Observable = require("./mixin/observable");

  WithAuthorization = require("./mixin/withAuthorization");

  Avatar = (function() {
    $.extend(Avatar.prototype, Observable, WithAuthorization);

    function Avatar(client, attributes) {
      this.client = client;
      this.setAttributes(attributes);
      if (this.notifiableEvents == null) {
        this.notifiableEvents = [];
      }
      this.notifiableEvents = this.notifiableEvents.concat(["build", "equip", "colour1", "colour2", "representationChange", "colourEnsemble"]);
      this.clear();
    }

    Avatar.prototype.representation = function(representation, notify) {
      if (notify == null) {
        notify = true;
      }
      if (representation != null) {
        this.metadata("representation", representation);
        if (notify) {
          this.notify("representationChange");
        }
      }
      return this.metadata("representation");
    };

    Avatar.prototype.clear = function() {
      return this._ensemble = {};
    };

    Avatar.prototype.build = function(ensemble) {
      var asset, category;
      this.clear();
      for (category in ensemble) {
        asset = ensemble[category];
        this._ensemble[category] = Lang.clone(asset);
      }
      return this.notify("build");
    };

    Avatar.prototype.equip = function(category, asset, colours) {
      if (colours == null) {
        colours = null;
      }
      asset = Lang.clone(asset);
      if (colours) {
        asset = this.setAssetColours(asset, colours);
      }
      this._ensemble[category] = asset;
      this.notify("equip", {
        category: category
      });
      return this._ensemble[category];
    };

    Avatar.prototype.equipped = function(category) {
      return this._ensemble[category];
    };

    Avatar.prototype.ensemble = function() {
      return Lang.clone(this._ensemble);
    };

    Avatar.prototype.colour1 = function(colour, category) {
      return this.setColour(colour, 1, category);
    };

    Avatar.prototype.colour2 = function(colour, category) {
      return this.setColour(colour, 2, category);
    };

    Avatar.prototype.colourEnsemble = function(palette) {
      var category, colour, colours, position, value;
      for (category in palette) {
        colours = palette[category];
        for (colour in colours) {
          value = colours[colour];
          position = colour.slice(0);
          this.setColourOnCategory(value, position, category);
        }
      }
      return this.notify("colourEnsemble");
    };

    Avatar.prototype.imageUrl = function(options) {
      var download;
      if (options == null) {
        options = {};
      }
      if (this.attributes.Id == null) {
        throw new Error("Avatar: must have an Id attribute to return an imageUrl.");
      }
      if (options.representation == null) {
        options.representation = this.representation();
      }
      if (options.representation == null) {
        throw new Error("Avatar: you must provide a representation when calling Avatar::imageUrl().");
      }
      download = options.download ? "&download=true" : "";
      return "/avatars/" + this.attributes.Id + "/image/" + options.representation + "?timestamp=" + (this.timestamp()) + download;
    };

    Avatar.prototype.save = function() {
      var d,
        _this = this;
      d = $.Deferred();
      return this.withAuthorization({
        deferred: d,
        action: function() {
          var key, payload, prelude, value, _ref;
          if (_this.attributes.User == null) {
            throw new Error("Avatar: can't save an Avatar without a User attribute");
          }
          _this.updateTimestamp();
          payload = {};
          _ref = _this.attributes;
          for (key in _ref) {
            value = _ref[key];
            payload[key] = value;
          }
          if (payload.Label == null) {
            payload.Label = _this.randomLabel();
          }
          prelude = payload.Id != null ? "update" : "create";
          return _this[prelude](payload).then(function() {
            return _this.saveAssets();
          }).then(function() {
            return d.resolve();
          });
        }
      });
    };

    Avatar.prototype["delete"] = function() {
      var d,
        _this = this;
      if (this.attributes.Id == null) {
        throw new Error("Avatar: can't delete an Avatar without an Id attribute.");
      }
      d = $.Deferred();
      this.withAuthorization({
        deferred: d,
        action: function() {
          return _this.client["delete"]("/avatars/" + _this.attributes.Id, function(err, res) {
            if (err) {
              return d.reject(err);
            }
            return d.resolve();
          });
        }
      });
      return d.promise();
    };

    Avatar.prototype.assetList = function() {
      var asset, category, list, _ref;
      list = [];
      _ref = this.ensemble();
      for (category in _ref) {
        asset = _ref[category];
        list.push({
          AssetId: asset.Id,
          Colour1: asset.Colour1 || null,
          Colour2: asset.Colour2 || null
        });
      }
      return list;
    };

    Avatar.prototype.metadata = function(key, value) {
      var metadata, _ref;
      if (key == null) {
        key = null;
      }
      if (key === null) {
        return this.parseMetadata();
      }
      if (typeof value === "undefined") {
        if (typeof key === "object") {
          return this.attributes.Metadata = JSON.stringify(key);
        }
        if (typeof key === "string") {
          return (_ref = this.parseMetadata()) != null ? _ref[key] : void 0;
        }
      }
      metadata = this.parseMetadata() || {};
      metadata[key] = value;
      this.attributes.Metadata = JSON.stringify(metadata);
      return value;
    };

    Avatar.prototype.parseMetadata = function() {
      var e;
      try {
        return JSON.parse(this.attributes.Metadata);
      } catch (_error) {
        e = _error;
        return null;
      }
    };

    Avatar.prototype.timestamp = function() {
      return this.metadata("timestamp") || this.updateTimestamp();
    };

    Avatar.prototype.updateTimestamp = function() {
      return this.metadata("timestamp", String(new Date().getTime()));
    };

    Avatar.prototype.setAssetColours = function(asset, colours) {
      asset.Colour1 = colours.colour1 || null;
      asset.Colour2 = colours.colour2 || null;
      return asset;
    };

    Avatar.prototype.setColour = function(colour, position, category) {
      if (category != null) {
        this.setColourOnCategory(colour, position, category);
        return this.notify("colour" + position, {
          category: category
        });
      } else {
        this.setColourGlobal(colour, position);
        return this.notify("colour" + position);
      }
    };

    Avatar.prototype.setColourOnCategory = function(colour, position, category) {
      var attribute;
      attribute = "Colour" + position;
      return this.equipped(category)[attribute] = colour;
    };

    Avatar.prototype.setColourGlobal = function(colour, position) {
      var category, _results;
      _results = [];
      for (category in this.ensemble()) {
        _results.push(this.setColourOnCategory(colour, position, category));
      }
      return _results;
    };

    Avatar.prototype.setAttributes = function(attributes) {
      var key, value, _results;
      if (this.attributes == null) {
        this.attributes = {};
      }
      _results = [];
      for (key in attributes) {
        value = attributes[key];
        _results.push(this.attributes[key] = value);
      }
      return _results;
    };

    Avatar.prototype.randomLabel = function() {
      return "hapi-avatar-" + (Math.random() * 1e18).toString(16);
    };

    Avatar.prototype.saveAssets = function() {
      var d, payload, url;
      d = $.Deferred();
      url = "/avatars/" + this.attributes.Id + "/assets";
      payload = this.truncateColours(this.assetList());
      this.client.post(url, payload, function(err) {
        if (err) {
          return d.reject(err);
        }
        return d.resolve();
      });
      return d.promise();
    };

    Avatar.prototype.truncateColours = function(assetList) {
      var asset, i, _i, _j, _len, _len1, _ref, _ref1;
      for (_i = 0, _len = assetList.length; _i < _len; _i++) {
        asset = assetList[_i];
        _ref = [1, 2];
        for (_j = 0, _len1 = _ref.length; _j < _len1; _j++) {
          i = _ref[_j];
          asset["Colour" + i] = ((_ref1 = asset["Colour" + i]) != null ? _ref1.slice(-6) : void 0) || null;
        }
      }
      return assetList;
    };

    Avatar.prototype.create = function(payload) {
      var d, url,
        _this = this;
      d = $.Deferred();
      url = "/users/" + this.attributes.User + "/avatars";
      this.client.post(url, payload, function(err, response) {
        if (err) {
          throw err;
        }
        _this.setAttributes(response);
        return d.resolve();
      });
      return d.promise();
    };

    Avatar.prototype.update = function(payload) {
      var d, url,
        _this = this;
      d = $.Deferred();
      url = "/avatars/" + payload.Id;
      this.client.put(url, payload, function(err, response) {
        if (err) {
          throw err;
        }
        _this.setAttributes(response);
        return d.resolve();
      });
      return d.promise();
    };

    Avatar.prototype.notifyObserver = function(observer, event, data) {
      if (data == null) {
        data = {};
      }
      data.avatar = this;
      return observer.notify(event, data);
    };

    Avatar.prototype.checkCategory = function(category) {
      if (this._ensemble[category] == null) {
        throw new Error("No such category: " + category);
      }
    };

    return Avatar;

  })();

  if (typeof module !== "undefined" && module !== null) {
    module.exports = Avatar;
  }

}).call(this);
