// totem/lib/drawing/compositor.js
// Generated by CoffeeScript 1.6.3
(function() {
  var Compositor;

  Compositor = (function() {
    function Compositor(categories) {
      this.categories = categories;
    }

    Compositor.prototype.prepare = function(ensemble) {
      var category, categoryStack, composite, _i, _len, _ref, _ref1;
      composite = [];
      _ref = this.categories;
      for (_i = 0, _len = _ref.length; _i < _len; _i++) {
        category = _ref[_i];
        categoryStack = {};
        if (((_ref1 = ensemble[category]) != null ? _ref1.Layers : void 0) != null) {
          categoryStack[category] = this.prepareStack(ensemble[category]);
          composite.push(categoryStack);
        }
      }
      return composite;
    };

    Compositor.prototype.prepareStack = function(asset) {
      var layer, stack, _i, _len, _ref;
      asset = this.normaliseAssetColours(asset);
      stack = [];
      _ref = asset.Layers;
      for (_i = 0, _len = _ref.length; _i < _len; _i++) {
        layer = _ref[_i];
        stack.push(this.prepareCell(layer, [asset.Colour1, asset.Colour2]));
      }
      return stack;
    };

    Compositor.prototype.prepareCell = function(layer, colours) {
      var cell, colour;
      cell = {
        image: layer.Uri,
        colour: null
      };
      if ((layer.Colour != null) > 0 && (colour = colours[layer.Colour - 1])) {
        cell.colour = colour;
      }
      return cell;
    };

    Compositor.prototype.normaliseColour = function(colour) {
      var re;
      re = /(\#|0x)?[0-9a-fA-F]{6}/;
      if (!re.test(colour)) {
        return null;
      }
      return "#" + colour.slice(-6);
    };

    Compositor.prototype.normaliseAssetColours = function(asset) {
      var c, i, _i, _len, _ref;
      _ref = [1, 2];
      for (_i = 0, _len = _ref.length; _i < _len; _i++) {
        i = _ref[_i];
        c = "Colour" + i;
        asset[c] = this.normaliseColour(asset[c]);
      }
      return asset;
    };

    return Compositor;

  })();

  if (typeof module !== "undefined" && module !== null) {
    module.exports = Compositor;
  }

}).call(this);
