// totem/lib/drawing/service.js
// Generated by CoffeeScript 1.6.3
(function() {
  var DrawingService, Observable;

  Observable = require("../mixin/observable");

  DrawingService = (function() {
    $.extend(DrawingService.prototype, Observable);

    DrawingService.prototype.notifiableEvents = ["startedDrawing", "finishedDrawing"];

    function DrawingService(compositor) {
      this.compositor = compositor;
    }

    DrawingService.prototype.avatarOnCanvas = function(canvas) {
      var canvasId, _ref, _ref1;
      canvasId = this.canvasIdFor(canvas);
      return (_ref = this.registry) != null ? (_ref1 = _ref[canvasId]) != null ? _ref1.avatar : void 0 : void 0;
    };

    DrawingService.prototype.allAvatars = function() {
      var canvasId, record, _ref, _results;
      if (this.registry == null) {
        return [];
      }
      _ref = this.registry;
      _results = [];
      for (canvasId in _ref) {
        record = _ref[canvasId];
        _results.push(record.avatar);
      }
      return _results;
    };

    DrawingService.prototype.startDrawing = function(avatar, canvas, options) {
      var canvasId;
      if (this.registry == null) {
        this.registry = {};
      }
      canvasId = this.canvasIdFor(canvas);
      this.registry[canvasId] = {
        avatar: avatar,
        canvas: canvas,
        adaptor: options.adaptor,
        autoDraw: options.autoDraw
      };
      return this.draw(avatar, canvas, options.adaptor);
    };

    DrawingService.prototype.autoDraw = function() {
      var canvasId, registered, _ref, _results;
      if (this.registry == null) {
        return;
      }
      _ref = this.registry;
      _results = [];
      for (canvasId in _ref) {
        registered = _ref[canvasId];
        if (registered.autoDraw) {
          _results.push(this.draw(registered.avatar, registered.canvas, registered.adaptor));
        }
      }
      return _results;
    };

    DrawingService.prototype.canvasIdFor = function(canvas) {
      if (!$(canvas).data("totem-id")) {
        $(canvas).data("totem-id", (Math.random() * 1e18).toString(16));
      }
      return $(canvas).data("totem-id");
    };

    DrawingService.prototype.draw = function(avatar, canvas, adaptor) {
      var composite, data,
        _this = this;
      data = {
        avatar: avatar,
        canvas: canvas
      };
      this.notify("startedDrawing", data);
      composite = this.compositor.prepare(avatar.ensemble());
      return adaptor.drawOn(composite, canvas).then(function() {
        return _this.notify("finishedDrawing", data);
      });
    };

    return DrawingService;

  })();

  if (typeof module !== "undefined" && module !== null) {
    module.exports = DrawingService;
  }

}).call(this);
